treat validator updates as set

ancazamfir opened the pull request:
<!-- Thanks for filing a PR! Before hitting the button, please check the following items.-->

* [ ] Updated all relevant documentation in docs
* [ ] Updated all code comments where relevant
* [ ] Wrote tests
* [ ] Updated CHANGELOG_PENDING.md

Liamsi commented on types/validator_set.go:
```diff
@@ -94,6 +97,20 @@ func (vals *ValidatorSet) IncrementProposerPriority(times int) {
+	tvp := vals.TotalVotingPower()
+	diff := computeMaxMinPriorityDiff(vals)
+	if diff > K * tvp {
+		ratio := float64(K * tvp) / float64(diff)
```
There were some concerns around on using floats (probably on slack, I'll see if I can dig those up).

ancazamfir replied:
Ok, please let me know. This will most likely be considered under https://github.com/tendermint/tendermint/pull/3182/
I just did some experimentation here. The main objective was to remove the diff from the parameters and contain the priority window size  here. Not sure the caller should decide on this.

ancazamfir replied:
removed

Liamsi commented on state/execution.go:
```diff
@@ -448,7 +383,8 @@ func updateState(
 	if len(validatorUpdates) > 0 {
-		err := updateValidators(nValSet, validatorUpdates)
+		//err := nValSet.UpdateWithChangeSet(validatorUpdates)
+		err := nValSet.UpdateWithChangeList(validatorUpdates)
```
Question: Is `UpdateWithChangeSet` only commented out because it would break some of the tests? This will eventually change to `UpdateWithChangeSet`, right?

ancazamfir replied:
I did did the change and fixed the test. It took a long time to fix these two in state_test.go
- TestProposerPriorityDoesNotGetResetToZero
- TestProposerPriorityProposerAlternates
Because the code now changes the penalty for newly added validators to updated total power (instead of previous total power) and because the centering is done before IncrementProposerPriority() it has been a bit of a nightmare to get the numbers right.
And in the end we manually match the numbers to the actual code we are testing which is not the best thing really. Ideally these tests should move to validator_set_test.go and instead of checking exact numbers check that centering and scaling achieved the desired results, that the selection is fair etc.

Liamsi replied:
> and instead of checking exact numbers check that centering and scaling achieved the desired results, that the selection is fair etc.

I agree. Since introduced (as requested in a review) it was a PITA to update those with every change (and there were a few since then). Maybe it is fine to have one such test as some kind of regression test which can easily updated.

Liamsi commented on types/validator_set.go:
```diff
@@ -366,6 +383,266 @@ func (vals *ValidatorSet) Iterate(fn func(index int, val *Validator) bool) {
+// If more or equal than 1/3 of total voting power changed in one block, then
+// a light client could never prove the transition externally. See
+// ./lite/doc.go for details on how a light client tracks validators.
+func (currentSet *ValidatorSet) UpdateWithChangeList(updates []*Validator) error {
```
Will `UpdateWithChangeList` become obsolete?

ancazamfir replied:
Not sure if we want to keep both APIs. Until we decide I leave it here.

Liamsi replied:
Suggestion: I would still remove it, when it is not used anywhere. If the need arises to treat updates as a list, we can readd it, or, open up a way for the application to control how updates are treated.

Liamsi commented on types/validator_set_test.go:
```diff
@@ -599,3 +599,217 @@ func TestValidatorSetVerifyCommit(t *testing.T) {
+	v4 = newValidator([]byte("v4"), 0)
+	delVals := []*Validator{v3, v4}
+	assert.Nil(t, vals.UpdateWithChangeSet(delVals))
+	assert.Nil(t, verifyValidatorSet(vals))
```
This looks like this could be changed to [table-driven tests](https://github.com/golang/go/wiki/TableDrivenTests).

ancazamfir replied:
Ok, let me try this.

ancazamfir replied:
changed as suggested

Liamsi commented on types/validator_set_test.go:
```diff
@@ -599,3 +599,217 @@ func TestValidatorSetVerifyCommit(t *testing.T) {
+
+	chVals = []*Validator{v5, v3}
+	assert.Nil(t, vals2.UpdateWithChangeSet(chVals))
+	assert.Nil(t, verifyValidatorSet(vals2))
```
This could be changed to: `assert.NoError`. In that way the reader immediately would notice, that  `verifyValidatorSet` returns an error.

Liamsi replied:
Alternatively (as this is a test), you could also pass around `t` and do the asserts inside of `verifyValidatorSet(t *testing.T, vals2)` etc.

ancazamfir replied:
Will try both suggestions, thanks!

Liamsi commented on types/validator_set.go:
```diff
@@ -366,6 +383,266 @@ func (vals *ValidatorSet) Iterate(fn func(index int, val *Validator) bool) {
+	prevAddr := []byte(nil)
+
+	for _, valUpdate := range changes {
+		if bytes.Equal(valUpdate.Address, prevAddr) {
```
Are the `changes` also expected to be sorted? If not, I do not think this would catch all duplicates. If they are in arbitrary order it is not enough to check against `prevAddr`.

Liamsi commented on types/validator_set_test.go:
```diff
@@ -599,3 +599,217 @@ func TestValidatorSetVerifyCommit(t *testing.T) {
+		nvs := make([]*Validator, m)
+		for j := 0; j < m; j++ {
+			nvs[j] = newValidator([]byte(fmt.Sprintf("v%d", j+len)), 1000)
+		}
```
If I understand this correctly, this aims to benchmark the execution of `vals.UpdateWithChangeSet(nvs)`. If so, you could add a line here: 
```suggestion
		}
		b.ResetTimer()
```
As everything until line 813 is initialization for this.

Also, you can remove the assert below (it probably uses reflection and might slow down the benchmark results).

Liamsi replied:
My bad. Resetting the timer inside of the loop will make the benchmarks take forever (as they do not "converge").

Liamsi replied:
The correct way to do this is probably to move the init code out of the "b.N-loop" (and wrap it with b.StopTimer & b.StartTimer)

ancazamfir replied:
I made the change, please take a look.

Liamsi commented on types/validator_set_test.go:
```diff
@@ -599,3 +599,217 @@ func TestValidatorSetVerifyCommit(t *testing.T) {
+		assert.Nil(b, verifyValidatorSet(vals))
+		len := len(vals.Validators)
+		// Add m validators
+		m := 2000
```
I do not think we currently care for more than a few hundred/max 1000 validators (?).

ancazamfir replied:
Understood, but it's good practice IMO to try high scale to understand what could go wrong. Also I experimented a bit with the go cpu profiler here :)

Liamsi commented on types/validator_set.go:
```diff
@@ -74,35 +77,57 @@ func (vals *ValidatorSet) CopyIncrementProposerPriority(times int) *ValidatorSet
 	// NOTE: This check is merely a sanity check which could be
 	// removed if all tests would init. voting power appropriately;
 	// i.e. diffMax should always be > 0
 	if diffMax <= 0 {
```
related: https://github.com/tendermint/tendermint/pull/3049#discussion_r251341684

ancazamfir replied:
Should we fix this in https://github.com/tendermint/tendermint/pull/3182?

Liamsi replied:
Yes, it's OK to this in  #3182 or in another PR.

Liamsi commented on state/execution.go:
```diff
@@ -448,7 +383,8 @@ func updateState(
 	lastHeightValsChanged := state.LastHeightValidatorsChanged
 	if len(validatorUpdates) > 0 {
-		err := updateValidators(nValSet, validatorUpdates)
+		//err := nValSet.UpdateWithChangeSet(validatorUpdates)
```
delete this line
```suggestion
	
```

Liamsi commented on types/validator_set.go:
```diff
@@ -366,6 +385,202 @@ func (vals *ValidatorSet) Iterate(fn func(index int, val *Validator) bool) {
+// Merges the vals' validator list with the updates list.
+// When two elements with same address are seen, the one from updates is selected.
+// Expects updates to be a list of updates sorted by address with no duplicates,
+// and to have been validated with validateUpdates().
```
I can't find `validateUpdates`. This probably refers to `verifyUpdatesAndComputeNewPriorities`?

Liamsi commented on types/validator_set.go:
```diff
@@ -366,6 +385,202 @@ func (vals *ValidatorSet) Iterate(fn func(index int, val *Validator) bool) {
+// total voting power would not exceed if these updates would be applied to the set.
+// Computes the proposer priority for the validators not present in the set.
+// Returns error if first step fails.
+// updates parameter must be a list of unique validators to be added or updated.
```
Maybe add a comment that explains that this modifies the priorities of the passed `updates` but does not change `vals`?

Liamsi commented on types/validator_set.go:
```diff
@@ -173,21 +185,28 @@ func (vals *ValidatorSet) getValWithMostPriority() *Validator {
+// Makes a copy of the validator list
+func validatorListCopy(valsList []*Validator) []*Validator {
+	valsCopy := make([]*Validator, len(valsList))
+	for i, val := range(valsList) {
```
```suggestion
	for i, val := range valsList {
```

Liamsi commented on types/validator_set.go:
```diff
@@ -366,6 +385,202 @@ func (vals *ValidatorSet) Iterate(fn func(index int, val *Validator) bool) {
+
+// Removes the validators specified in 'deletes' from validator set 'vals'.
+// Should not fail as verification has been done before.
+func (vals *ValidatorSet)applyRemovals(deletes []*Validator) {
```
Formatting seems off:
```suggestion
func (vals *ValidatorSet) applyRemovals(deletes []*Validator) {
```

Liamsi replied:
You could add goimports or gofmt to your workflow/IDE to have the code formatted in an idiomatic way.

ancazamfir replied:
Ouch...I just switched to trying GoLand (from VSCode) and forgot to add gofmt to the file watchers...thanks!

Liamsi commented on types/validator_set.go:
```diff
@@ -366,6 +385,202 @@ func (vals *ValidatorSet) Iterate(fn func(index int, val *Validator) bool) {
+	vals.applyRemovals(deletes)
+	vals.RescalePriorities(K * vals.TotalVotingPower())
+	vals.shiftByAvgProposerPriority()
+
```
Question: If I understand correctly, the methods `ValidatorSet.Add`, `ValidatorSet.Update` and  `ValidatorSet.Remove` should not be used anywhere, or, only in tests. We probably still want to expose them? Or should we make them unexported?

ancazamfir replied:
I removed them and changed the tests. Also I learnt today from @melekes about reflect.DeepEqual() so I changed the tests to use that.

Liamsi replied:
cool! I think, `assert.Equal` uses `reflect.DeepEqual()` internally, too.

Liamsi commented on types/validator_set.go:
```diff
@@ -366,6 +385,202 @@ func (vals *ValidatorSet) Iterate(fn func(index int, val *Validator) bool) {
+// Expects updates to be a list of updates sorted by address with no duplicates,
+// and to have been validated with validateUpdates().
+// The validator's priorities in 'updates' should be already computed.
+func (vals *ValidatorSet) applyUpdates(updates []*Validator) {
```
Even if this is a private method: can we have a simple test that show-cases how a given ValidatorSet.Validators is expected to look after applyUpdates?

ancazamfir replied:
added tests

ancazamfir commented on types/validator_set.go:
```diff
@@ -366,6 +383,266 @@ func (vals *ValidatorSet) Iterate(fn func(index int, val *Validator) bool) {
+// - applies the removals against the validator set and performs scaling and centering of priority values
+// If error is detected during verification steps an error is returned and the validator set
+// is not changed.
+func (vals *ValidatorSet) UpdateWithChangeSet(changes []*Validator) error {
```
:)

Liamsi commented on state/state_test.go:
```diff
@@ -322,7 +322,8 @@ func TestProposerFrequency(t *testing.T) {
 		for j := 0; j < N; j++ {
-			votePower := int64(cmn.RandInt() % maxPower)
+			// make sure votePower > 0
+			votePower := int64(cmn.RandInt()%maxPower) + 1
```
Can't this (rarely) exceed maxPower now?

ancazamfir replied:
yes, and I think with +1 we can get 1000 but no more.

Liamsi replied:
You are right! Sorry for the noise.

Liamsi commented on types/validator_set_test.go:
```diff
@@ -602,195 +602,316 @@ func TestValidatorSetVerifyCommit(t *testing.T) {
-}
-
-func getTotalProposerPriority(vals *ValidatorSet) int64 {
+type TestVal struct {
```
Could be `testVal` (unexported)

Liamsi commented on types/validator_set.go:
```diff
@@ -42,19 +45,17 @@ type ValidatorSet struct {
-	vals := &ValidatorSet{
-		Validators: validators,
+	vals := &ValidatorSet{}
+	err := vals.UpdateWithChangeSet(valz)
```
It's interesting that this doesn't change the behaviour now. IIUC, calling `NewValidatorSet` would initialize priorities with one call of `incrementProposerPriority` (basically setting all prios to the voting power of that validator and decrement the mostest). Now, we detect duplicates and additionally set all initial prios to -1.125*total (then rescale, shift by average via `UpdateWithChangeSet`) and then do the `incrementProposerPriority` (which rescales / shifts again).
The test suggest this doesn't have any side effects. And as we are not constantly calling `NewValidatorSet`, I think it is negligable to do the rescaling/shifting several times here.

Liamsi replied:
If we do not want that, and to catch the duplicates, we could call the private helper functions (processChanges, verifyUpdates) here directly (would duplicate some of the code of `UpdateWithChangeSet`)

ancazamfir replied:
> It's interesting that this doesn't change the behaviour now. IIUC, calling `NewValidatorSet` would initialize priorities with one call of `incrementProposerPriority` (basically setting all prios to the voting power of that validator and decrement the mostest). Now, we detect duplicates and additionally set all initial prios to -1.125*total (then rescale, shift by average via `UpdateWithChangeSet`) and then do the `incrementProposerPriority` (which rescales / shifts again).

Yes true it's an overhead. But I think we will remove the rescale from incrementProposerPriority() once we know the priority window bound. 

> The tests suggest this doesn't have any side effects. And as we are not constantly calling `NewValidatorSet`, I think it is negligable to do the rescaling/shifting several times here.

Yes, this is what creates the validator set so only called once. After this we have the updates that do pretty much the same.

>If we do not want that, and to catch the duplicates, we could call the private helper functions (processChanges, verifyUpdates) here directly (would duplicate some of the code of `UpdateWithChangeSet`)

My concern is not really the extra rescaling/ centering. That I think will get better as I mentioned above. The problem right now is the different semantics of the parameter: - --
- NewValidatorSet() : add all validators in list, vs.
- UpdateWithChangeSet(): if validator exist either update or delete, if not add

So right now, when we call NewValidatorSet() with a validator with power 0 we fail (as we should) but the error says something like "cannot delete validator with address blah.."
I will fix this soon.

ancazamfir replied:
Made some changes so we have better errors.

Liamsi commented on types/validator_set.go:
```diff
@@ -366,6 +335,204 @@ func (vals *ValidatorSet) Iterate(fn func(index int, val *Validator) bool) {
+// Returns error if first step fails.
+// 'updates' parameter must be a list of unique validators to be added or updated.
+// No changes are made to the validator set 'vals'.
+func verifyUpdatesAndComputeNewPriorities(updates []*Validator, vals *ValidatorSet) error {
```
How do you think about splitting this into two methods? `verifyUpdates` which only does the first loop and `computePriorities` which does the 2nd?

ancazamfir replied:
I did have the same thought but didn't want to carry over the info from verify into compute.
I realize that Go programmers really like small small functions :) Sooo...I will split

ancazamfir replied:
split and rearranged.

milosevic commented on state/execution.go:
```diff
@@ -448,7 +383,7 @@ func updateState(
 	lastHeightValsChanged := state.LastHeightValidatorsChanged
 	if len(validatorUpdates) > 0 {
-		err := updateValidators(nValSet, validatorUpdates)
+		err := nValSet.UpdateWithChangeSet(validatorUpdates)
```
Like the name.

milosevic commented on state/state_test.go:
```diff
@@ -424,49 +425,71 @@ func TestProposerPriorityDoesNotGetResetToZero(t *testing.T) {
+	wantVal1Prio += updatedVal1.VotingPower // 1
+	wantVal2Prio -= totalPower              // -1
+
+	assert.Equal(t, wantVal2Prio, updatedVal2.ProposerPriority)
```
This test is very big. Can it be split into several more focused one?

ancazamfir replied:
yes, my nightmare on this pull. In one version I had most of the last part removed.

milosevic commented on types/validator_set.go:
```diff
@@ -19,7 +19,10 @@ import (
-const MaxTotalVotingPower = int64(math.MaxInt64) / 8
+const (
+	MaxTotalVotingPower = int64(math.MaxInt64) / 8
+	K                   = 2
```
What is K?

ancazamfir replied:
it's 2 :)...changed to a nice long name and added comment explaining.

milosevic commented on types/validator_set.go:
```diff
@@ -42,19 +45,17 @@ type ValidatorSet struct {
 // NewValidatorSet initializes a ValidatorSet by copying over the
 // values from `valz`, a list of Validators. If valz is nil or empty,
 // the new ValidatorSet will have an empty list of Validators.
+// The addresses of vlaidators in `valz` must be unique otherwise the
```
Typo. vlaidators -> validators.

milosevic commented on types/validator_set.go:
```diff
@@ -366,6 +335,204 @@ func (vals *ValidatorSet) Iterate(fn func(index int, val *Validator) bool) {
 }
 
+// Checks changes against duplicates, splits the changes in updates and removals, sorts them by address
+func processChanges(origChanges []*Validator) (updates, removals []*Validator, err error) {
```
Maybe changes can be split into additions, removals and updates. The potential benefit is the fact that we can easily determine size of the new valset, and we can probably also get rid of `Remove` function (like we get rid of `Add`) function.

ancazamfir replied:
I added an applyRemove2() function to show how it would look without Remove().

ebuchman replied:
Is `split into additions, removals and updates` related to `also get rid of Remove function` ? 

Also you guys inspired me to write the shorted applyRemove I could. Check it out below!

ebuchman replied:
> split into additions, removals and updates

This sounds good, but it might make things more complicated in `updateWithChangeSet` since we have to do a few things for both additions and updates. Not sure what's best.

milosevic replied:
We probably don't need to separate updates further, but I have other proposal. What if instead of `applyUpdates` and `applyRemovals` we have single function that just applies updates and computes nextValidatorSet. This function would start from empty val set (size is equal to existing.size + additions.size - deletions.size) and then iterate over existing valset and for each validator check if it is update, addition or deletion, and then do appropriate logic. For deletions we don't do anything; for updates and additions, we add it to the new set and compute proposer priority (in place; I am not sure `computeNewPriorites` function brings a lot). To make this possible, updates and deletions should be maps, and `processChanges` would need to return size of the new val set. Verify methods would just return error and not compute modified voting power as they do in the last version, so it might be a bit cleaner. `Remove` method is not needed anymore.

milosevic commented on types/validator_set.go:
```diff
@@ -366,6 +335,204 @@ func (vals *ValidatorSet) Iterate(fn func(index int, val *Validator) bool) {
+// - applies the removals against the validator set and performs scaling and centering of priority values
+// If error is detected during verification steps an error is returned and the validator set
+// is not changed.
+func (vals *ValidatorSet) UpdateWithChangeSet(changes []*Validator) error {
```
Beautiful code refactoring. I just have question regarding contract we have with ABCI app, i.e., what happens if an application returns changes which contains duplicates for example?

ancazamfir replied:
Yes, the error bubbles up:
UpdateWithChangeSet() -> execution.go:updateState() -> execution.go:ApplyBlock() -> reactor.go:poolRoutine() where we panic

We could probably discard the updates, notify the application and continue without the validator set changes.
It is also possible to discard duplicates and invalid changes and apply the rest good ones...do we want to do this?

ebuchman replied:
Since we don't have a good way to "notify the application" right now, and since something like duplicate validators probably indicates a serious error in application logic, we should probably treat it as is and have tendermint halt. Of course this means we need to better support rollbacks for devs with buggy apps ... #3196

ancazamfir replied:
Also to @milosevic earlier point, should I update the ABCI documentation? Part of this PR? Or someone else needs/ wants to do it?

ebuchman replied:
Can be done separately, good point though.

ebuchman commented on types/validator_set.go:
```diff
@@ -366,6 +338,260 @@ func (vals *ValidatorSet) Iterate(fn func(index int, val *Validator) bool) {
+	merged := make([]*Validator, len(existing)-len(deletes))
+	i := 0
+
+	for len(existing) > 0 && len(deletes) > 0 {
```
What do you think of this:

    func popVal(vals []*Validator) []*Validator{
        if len(vals) > 1 { return vals[1:] }
        return vals
    }

    // ...

    for i := 0; i < len(merged); i++ {
		for existing[0] == deletes[0] {
			existing, deletes = popVal(existing), popVal(deletes)
		}
		merged[i], existing = existing[0], popVal(existing)
	}

🐙 

I think we could do something similar for applyUpdates if we separated add from update

melekes replied:
Why do all this? For speed? If yes, I'd rather prefer safer but slower version.

ebuchman commented on types/validator_set.go:
```diff
@@ -366,6 +338,260 @@ func (vals *ValidatorSet) Iterate(fn func(index int, val *Validator) bool) {
+	}
+
+	// Verify that applying the 'updates' against 'vals' will not result in error.
+	updatedTotalVotingPower, err := verifyUpdates(updates, vals)
```
don't we need to include changes from deletes in the updatedTotalVotingPower ?

ancazamfir replied:
Ouch!!! Totally missed this! Thanks, thanks!!!!

ancazamfir replied:
I will write a test for this soon

ancazamfir replied:
Hmm, actually it was correct before. Since we start with totalVotingPower and add the deltas for changes and power for adds.
But this proves that I am definitely missing explanations...and the test.

ebuchman replied:
But we don't subtract the deletes?

milosevic replied:
It seems I also misunderstand something. Shouldn't priority of new validator be equal to -1.25*totalVotingPower of the validator set before we apply any changes?

ancazamfir replied:
No we don't, it's on purpose. So the idea is to forget the deletes initially. And determine how the set would look like without them. This way the relative priority of the existing validators is kept and we make sure we add the new validators with smallest priority (-TVP' of this imaginary set). This keeps things fair across this set.
Then we delete the required validators, scale and center. When we do that we do not change the relative priority of remaining validators and we make sure the prio window stays centered and within the 2*TVP
We discussed this with @cwgoes and @ismail in https://github.com/tendermint/tendermint/issues/3181.

ancazamfir replied:
@milosevic - yes it used to be like that and I documented it thinking it was intentional. However, when I brought this up to @cwgoes and @Liamsi they mentioned that it is in fact a bug. So I changed it here...doc still needs to be updated.

ebuchman commented on types/validator_set.go:
```diff
@@ -366,6 +338,260 @@ func (vals *ValidatorSet) Iterate(fn func(index int, val *Validator) bool) {
+	//vals.applyRemovals2(deletes)
+
+	// Scale and center
+	vals.RescalePriorities(PriorityWindowSizeFactor * vals.TotalVotingPower())
```
Do we need to do this here? We call IncrementProposerPriority after UpdateWithChangeSet so this sequence would be happening twice.

ancazamfir replied:
Yes we could. But if we find the limit on prio distance then we can remove it from IncrementProposerPriority() and only rescale and center after changes

ebuchman replied:
Oh that's great. I guess it makes sense to leave it in both for now. Calling it twice in a row has the same effect as calling it once, right?

ancazamfir replied:
You mean calling it back-to-back? Yes, second time is a noop.

ebuchman commented:
I merged in develop to fix a build issue. But linter failing:

```
--> Running linter
types/validator_set_test.go:899:2: Consider preallocating `valList` (prealloc)
	var valList []*Validator
	^
state/state_test.go:486:47: `substract` is a misspelling of `subtract` (misspell)
	// v1(10):-9+10, v2(1):9+1 -> v2 proposer so substract tvp(11)
	                                             ^
types/validator_set.go:437:2: redundant return statement (gosimple)
	return
	^
types/validator_set.go:510:27: func `(*ValidatorSet).applyRemovals2` is unused (unused)
func (vals *ValidatorSet) applyRemovals2(deletes []*Validator) {
                          ^
Makefile:247: recipe for target 'lint' failed
```

Also some of my review comments are responses to previous conversations, which can be confusing if you don't realize. I guess we should try to avoid putting responses inside a new review.

melekes commented on types/validator_set.go:
```diff
@@ -366,6 +338,260 @@ func (vals *ValidatorSet) Iterate(fn func(index int, val *Validator) bool) {
+	merged := make([]*Validator, len(existing)+len(updates))
+	i := 0
+
+	for len(existing) > 0 && len(updates) > 0 {
```
a bit hard to follow. can't we just range over `updates`?

melekes commented on types/validator_set.go:
```diff
@@ -366,6 +338,260 @@ func (vals *ValidatorSet) Iterate(fn func(index int, val *Validator) bool) {
+	}
+
+	vals.Validators = merged[:i]
+	vals.totalVotingPower = 0
```
why reset `totalVotingPower`? Function declaration does not say anything about that

milosevic replied:
This is probably leftover of old implementation. But we should probably avoid doing it and just always set it to a correct value.

ancazamfir replied:
@melekes @milosevic ...yes I will look into it
valSet.TotalVotingPower() just returns valSet.totalVotingPower if it is non-zero. It should probably be enough to call valSet.TotalVotingPower() at the end of updateWithChangeSet()

melekes commented on types/validator_set.go:
```diff
@@ -366,6 +338,260 @@ func (vals *ValidatorSet) Iterate(fn func(index int, val *Validator) bool) {
+		i++
+	}
+	vals.Validators = merged[:i]
+	vals.totalVotingPower = 0
```
same question

melekes commented on types/validator_set_test.go:
```diff
@@ -599,3 +599,365 @@ func TestValidatorSetVerifyCommit(t *testing.T) {
+	return valSet
+}
+
+func getTotalProposerPriority(valSet *ValidatorSet) int64 {
```
This function should probably go into ValidatorSet itself. I.e. `ValidatorSet#TotalProposerPriority()`

ancazamfir commented on types/validator_set.go:
```diff
@@ -42,19 +48,17 @@ type ValidatorSet struct {
+	vals := &ValidatorSet{}
+	err := vals.updateWithChangeSet(valz, false)
+	if err != nil {
+		panic(fmt.Sprintf("cannot create validator set: %s", err))
```
I prefer return error but this being a bigger change I wanted to hear from you.

ancazamfir commented:
@ebuchman and @melekes - thanks for the review!!!
There are still a few pending issues but before I move on them, let's agree on the error handling strategy.
Now that we control all operations in validator_set, what about:
- if NewValidatorSet(validators) results in an error, we return to caller and let it deal with it. Currently somewhere up the stack we panic...or if we go with the changes in https://github.com/tendermint/tendermint/pull/3241, we exit. In the future the caller might do something else
(Note: I haven't done this yet, just added panic for now since this is a bigger change I wanted to confirm with you)
- if UpdateWithChangeSet(changes) results in error same as above. I added check that final set is not empty
- if somehow we are called with any valSet.method() with empty set, we panic since, given the above we should never run in this situation.

ancazamfir commented on types/validator_set.go:
```diff
@@ -366,6 +338,260 @@ func (vals *ValidatorSet) Iterate(fn func(index int, val *Validator) bool) {
+// updatedTotalVotingPower - the new total voting power if these updates would be applied
+// err - non-nil if the maximum allowed total voting power would be exceeded
+//
+// No changes are made to the validator set 'vals'.
```
yes

ebuchman commented:
The error handling suggestions sound good. I think they can be done in a separate PR though so we can get this merged.

ancazamfir commented:
Ok, I just pushed the fix for total voting power.

melekes commented on state/state_test.go:
```diff
@@ -483,7 +483,7 @@ func TestProposerPriorityDoesNotGetResetToZero(t *testing.T) {
 	// 3. Center - noop
 	// 4. IncrementProposerPriority() ->
-	// v1(10):-9+10, v2(1):9+1 -> v2 proposer so substract tvp(11)
+	// v1(10):-9+10, v2(1):9+1 -> v2 proposer so subsract tvp(11)
```
did you mean subtract?
